// jparticles/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.utils = exports.Events = exports.easing = exports.commonConfig = exports.WaveLoading = exports.Wave = exports.Snow = exports.Particle = exports.Line = void 0;
var config_1 = __importDefault(require("./common/config"));
exports.commonConfig = config_1.default;
var easing_1 = __importDefault(require("./common/easing"));
exports.easing = easing_1.default;
var events_1 = __importDefault(require("./common/events"));
exports.Events = events_1.default;
var utils = __importStar(require("./utils/index"));
exports.utils = utils;
var line_1 = __importDefault(require("./line"));
exports.Line = line_1.default;
var particle_1 = __importDefault(require("./particle"));
exports.Particle = particle_1.default;
var snow_1 = __importDefault(require("./snow"));
exports.Snow = snow_1.default;
var wave_1 = __importDefault(require("./wave"));
exports.Wave = wave_1.default;
var wave_loading_1 = __importDefault(require("./wave-loading"));
exports.WaveLoading = wave_loading_1.default;
exports.default = {
    Line: line_1.default,
    Particle: particle_1.default,
    Snow: snow_1.default,
    Wave: wave_1.default,
    WaveLoading: wave_loading_1.default,
    commonConfig: config_1.default,
    easing: easing_1.default,
    Events: events_1.default,
    utils: utils,
};
